File: app/src/main/java/com/ivanovsky/passnotes/data/repository/keepass/KeepassDatabaseRepository.java
package com.ivanovsky.passnotes.data.repository.keepass;

import static com.ivanovsky.passnotes.data.entity.OperationError.MESSAGE_FAILED_TO_OPEN_DEFAULT_DB_FILE;
import static com.ivanovsky.passnotes.data.entity.OperationError.newGenericError;

import android.content.Context;

import androidx.annotation.NonNull;

import com.ivanovsky.passnotes.data.ObserverBus;
import com.ivanovsky.passnotes.data.entity.FileDescriptor;
import com.ivanovsky.passnotes.data.entity.OperationResult;
import com.ivanovsky.passnotes.data.repository.GroupRepository;
import com.ivanovsky.passnotes.data.repository.GroupRepositoryWrapper;
import com.ivanovsky.passnotes.data.repository.NoteRepository;
import com.ivanovsky.passnotes.data.repository.NoteRepositoryWrapper;
import com.ivanovsky.passnotes.data.repository.TemplateRepository;
import com.ivanovsky.passnotes.data.repository.TemplateRepositoryWrapper;
import com.ivanovsky.passnotes.data.repository.encdb.EncryptedDatabase;
import com.ivanovsky.passnotes.data.repository.encdb.EncryptedDatabaseKey;
import com.ivanovsky.passnotes.data.repository.EncryptedDatabaseRepository;
import com.ivanovsky.passnotes.data.repository.file.FSOptions;
import com.ivanovsky.passnotes.data.repository.file.FileSystemProvider;
import com.ivanovsky.passnotes.data.repository.file.FileSystemResolver;
import com.ivanovsky.passnotes.data.repository.file.OnConflictStrategy;
import com.ivanovsky.passnotes.domain.DatabaseLockInteractor;
import com.ivanovsky.passnotes.domain.entity.DatabaseStatus;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;

import timber.log.Timber;

public class KeepassDatabaseRepository implements EncryptedDatabaseRepository {

    private static final String EMPTY_DB_PATH = "base.kdbx.xml";

    private volatile KeepassDatabase db;
    @NonNull
    private final TemplateRepositoryWrapper templateRepositoryWrapper;
    @NonNull
    private final GroupRepositoryWrapper groupRepositoryWrapper;
    @NonNull
    private final NoteRepositoryWrapper noteRepositoryWrapper;
    private final Context context;
    private final FileSystemResolver fileSystemResolver;
    private final DatabaseLockInteractor lockInteractor;
    private final ObserverBus observerBus;
    private final Object lock;

    public KeepassDatabaseRepository(Context context,
                                     FileSystemResolver fileSystemResolver,
                                     DatabaseLockInteractor lockInteractor,
                                     ObserverBus observerBus) {
        this.context = context;
        this.fileSystemResolver = fileSystemResolver;
        this.lockInteractor = lockInteractor;
        this.observerBus = observerBus;
        this.lock = new Object();
        templateRepositoryWrapper = new TemplateRepositoryWrapper(this);
        groupRepositoryWrapper = new GroupRepositoryWrapper(this);
        noteRepositoryWrapper = new NoteRepositoryWrapper(this);
    }

    @Override
    public boolean isOpened() {
        return db != null;
    }

    @Override
    public EncryptedDatabase getDatabase() {
        return db;
    }

    @Override
    public NoteRepository getNoteRepository() {
        return noteRepositoryWrapper;
    }

    @Override
    public GroupRepository getGroupRepository() {
        return groupRepositoryWrapper;
    }

    @NonNull
    @Override
    public TemplateRepository getTemplateRepository() {
        return templateRepositoryWrapper;
    }

    @NonNull
    @Override
    public OperationResult<EncryptedDatabase> open(@NonNull EncryptedDatabaseKey key,
                                                   @NonNull FileDescriptor file,
                                                   @NonNull FSOptions options) {
        OperationResult<EncryptedDatabase> result;

        FileSystemProvider fsProvider = fileSystemResolver.resolveProvider(file.getFsAuthority());

        synchronized (lock) {
            if (db != null) {
                close();
            }

            OperationResult<InputStream> inputResult = fsProvider.openFileForRead(file,
                    OnConflictStrategy.CANCEL,
                    options);
            if (inputResult.isFailed()) {
                return inputResult.takeError();
            }

            OperationResult<KeepassDatabase> openResult = KeepassDatabase.open(fsProvider,
                    options,
                    file,
                    inputResult,
                    key,
                    observerBus::notifyDatabaseStatusChanged);
            if (openResult.isFailed()) {
                return openResult.takeError();
            }

            db = openResult.getObj();

            result = openResult.takeStatusWith(db);
        }

        if (db != null) {
            onDatabaseOpened(options, db.getStatus());
        }

        return result;
    }

    @NonNull
    @Override
    public OperationResult<Boolean> createNew(@NonNull EncryptedDatabaseKey key,
                                              @NonNull FileDescriptor file,
                                              boolean addTemplates) {
        OperationResult<Boolean> result;

        synchronized (lock) {
            FileSystemProvider fsProvider = fileSystemResolver.resolveProvider(file.getFsAuthority());

            OperationResult<InputStream> input;
            try {
                // 'in' will be closed in KeepassDatabase.open()
                InputStream in = new BufferedInputStream(context.getAssets().open(EMPTY_DB_PATH));
                input = OperationResult.success(in);
            } catch (IOException e) {
                Timber.d(e);
                return OperationResult.error(newGenericError(MESSAGE_FAILED_TO_OPEN_DEFAULT_DB_FILE));
            }

            EncryptedDatabaseKey unencryptedKey = new DefaultDatabaseKey();
            OperationResult<KeepassDatabase> openResult = KeepassDatabase.open(fsProvider,
                    FSOptions.defaultOptions(),
                    file,
                    input,
                    unencryptedKey,
                    null);
            if (openResult.isFailed()) {
                return openResult.takeError();
            }

            KeepassDatabase db = openResult.getObj();
            if (addTemplates) {
                OperationResult<Boolean> addTemplateResult = db.getTemplateRepository()
                        .addTemplates(TemplateFactory.createDefaultTemplates(), false);
                if (addTemplateResult.isFailed()) {
                    return addTemplateResult.takeError();
                }
            }

            // 'changeKey' will invoke commit
            OperationResult<Boolean> changeKeyResult = db.changeKey(unencryptedKey, key);
            if (changeKeyResult.isFailed()) {
                return changeKeyResult.takeError();
            }

            result = changeKeyResult;
        }

        return result;
    }

    @NonNull
    @Override
    public OperationResult<Boolean> close() {
        synchronized (lock) {
            if (db != null) {
                db = null;
            }
        }

        onDatabaseClosed();

        return OperationResult.success(true);
    }

    private void onDatabaseOpened(FSOptions fsOptions, DatabaseStatus status) {
        lockInteractor.onDatabaseOpened(fsOptions, status);
        observerBus.notifyDatabaseOpened(fsOptions, status);
    }

    private void onDatabaseClosed() {
        lockInteractor.onDatabaseClosed();
        observerBus.notifyDatabaseClosed();
    }
}

File: app/src/main/kotlin/com/ivanovsky/passnotes/data/repository/keepass/KeepassDatabaseRepository.kt
package com.ivanovsky.passnotes.data.repository.keepass

import com.ivanovsky.passnotes.data.ObserverBus
import com.ivanovsky.passnotes.data.entity.FileDescriptor
import com.ivanovsky.passnotes.data.entity.OperationError
import com.ivanovsky.passnotes.data.entity.OperationError.MESSAGE_FAILED_TO_GET_DATABASE
import com.ivanovsky.passnotes.data.entity.OperationError.newDbError
import com.ivanovsky.passnotes.data.entity.OperationResult
import com.ivanovsky.passnotes.data.repository.EncryptedDatabaseRepository
import com.ivanovsky.passnotes.data.repository.encdb.EncryptedDatabase
import com.ivanovsky.passnotes.data.repository.encdb.EncryptedDatabaseKey
import com.ivanovsky.passnotes.data.repository.file.FSOptions
import com.ivanovsky.passnotes.data.repository.file.FSOptions.Companion.defaultOptions
import com.ivanovsky.passnotes.data.repository.file.FileSystemResolver
import com.ivanovsky.passnotes.data.repository.file.OnConflictStrategy
import com.ivanovsky.passnotes.data.repository.keepass.keepassrs.KeepassRsDatabase
import com.ivanovsky.passnotes.data.repository.keepass.kotpass.KotpassDatabase
import com.ivanovsky.passnotes.domain.DatabaseLockInteractor
import com.ivanovsky.passnotes.domain.entity.exception.Stacktrace
import com.ivanovsky.passnotes.extensions.getOrThrow
import com.ivanovsky.passnotes.extensions.mapError
import com.ivanovsky.passnotes.extensions.mapWithObject
import com.ivanovsky.passnotes.util.toOperationResult
import java.io.InputStream
import java.util.concurrent.atomic.AtomicReference
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class KeepassDatabaseRepository(
    private val fileSystemResolver: FileSystemResolver,
    private val lockInteractor: DatabaseLockInteractor,
    private val syncStatusProvider: DatabaseSyncStateProvider,
    private val observerBus: ObserverBus
) : EncryptedDatabaseRepository {

    private val databaseRef = AtomicReference<DatabaseReference>()
    private val lock = ReentrantLock()

    override fun isOpened(): Boolean = (databaseRef.get() != null)

    override fun getDatabase(): EncryptedDatabase? = databaseRef.get()?.database

    override fun open(
        type: KeepassImplementation,
        key: EncryptedDatabaseKey,
        file: FileDescriptor,
        options: FSOptions
    ): OperationResult<EncryptedDatabase> {
        val fsProvider = fileSystemResolver.resolveProvider(file.fsAuthority)

        val openDbResult = lock.withLock {
            if (isOpened()) {
                close()
            }

            val openFileResult = fsProvider.openFileForRead(
                file,
                OnConflictStrategy.CANCEL,
                options
            )
            if (openFileResult.isFailed) {
                return@withLock openFileResult.takeError()
            }

            val openResult = openDatabase(
                type,
                fileSystemResolver,
                options,
                file,
                openFileResult,
                key
            )
            if (openResult.isFailed) {
                return@withLock openResult.takeError()
            }

            val db = openResult.obj
            databaseRef.set(DatabaseReference(type, db))
            openResult.takeStatusWith(db)
        }

        if (openDbResult.isSucceededOrDeferred) {
            onDatabaseOpened(openDbResult.getOrThrow(), openDbResult)
        }

        return openDbResult
    }

    override fun read(
        type: KeepassImplementation,
        key: EncryptedDatabaseKey,
        file: FileDescriptor,
        options: FSOptions
    ): OperationResult<EncryptedDatabase> {
        val fsProvider = fileSystemResolver.resolveProvider(file.fsAuthority)

        val reloadFileResult = fsProvider.getFile(file.path, options)
        if (reloadFileResult.isFailed) {
            return reloadFileResult.mapError()
        }

        val reloadedFile = reloadFileResult.getOrThrow()

        val openFileResult = fsProvider.openFileForRead(
            reloadedFile,
            OnConflictStrategy.CANCEL,
            options
        )
        if (openFileResult.isFailed) {
            return openFileResult.mapError()
        }

        return openDatabase(
            type,
            fileSystemResolver,
            options,
            reloadedFile,
            openFileResult,
            key
        )
    }

    override fun reload(): OperationResult<Boolean> {
        val result = lock.withLock {
            if (!isOpened()) {
                return@withLock OperationResult.error(
                    newDbError(
                        MESSAGE_FAILED_TO_GET_DATABASE,
                        Stacktrace()
                    )
                )
            }

            val oldDb = databaseRef.get().database
            val type = databaseRef.get().type
            val fsProvider = fileSystemResolver.resolveProvider(oldDb.getFile().fsAuthority)
            val fsOptions = oldDb.getFSOptions()
            val file = oldDb.getFile()
            val key = oldDb.getKey()

            val openFileResult = fsProvider.openFileForRead(
                file,
                OnConflictStrategy.CANCEL,
                fsOptions
            )
            if (openFileResult.isFailed) {
                return@withLock openFileResult.takeError()
            }

            val openResult = openDatabase(
                type = type,
                fsResolver = fileSystemResolver,
                fsOptions = fsOptions,
                file = file,
                input = openFileResult,
                key = key
            )
            if (openResult.isFailed) {
                return@withLock openResult.takeError()
            }

            val db = openResult.obj
            databaseRef.set(DatabaseReference(type, db))
            openResult.takeStatusWith(db)
        }

        if (result.isSucceededOrDeferred) {
            observerBus.notifyDatabaseDataSetChanged()
        }

        return result.mapWithObject(true)
    }

    override fun createNew(
        type: KeepassImplementation,
        key: EncryptedDatabaseKey,
        file: FileDescriptor,
        addTemplates: Boolean
    ): OperationResult<Boolean> {
        return lock.withLock {
            val dbResult = when (type) {
                KeepassImplementation.KOTPASS -> KotpassDatabase.new(
                    fsResolver = fileSystemResolver,
                    fsOptions = defaultOptions(),
                    file = file,
                    key = key,
                    isAddTemplates = addTemplates
                )

                KeepassImplementation.KEEPASS_RS -> OperationResult.error(
                    OperationError.newDbError(
                        OperationError.MESSAGE_WRITE_OPERATION_IS_NOT_SUPPORTED,
                        Stacktrace()
                    )
                )
            }
            if (dbResult.isFailed) {
                return@withLock dbResult.takeError()
            }

            OperationResult.success(true)
        }
    }

    override fun close(): OperationResult<Boolean> {
        lock.withLock {
            if (isOpened()) {
                databaseRef.get().database
                    .apply {
                        this.watcher.unsubscribe(syncStatusProvider)
                    }

                databaseRef.set(null)
            }
        }

        onDatabaseClosed()

        return OperationResult.success(true)
    }

    private fun onDatabaseOpened(
        db: EncryptedDatabase,
        openResult: OperationResult<EncryptedDatabase>
    ) {
        lockInteractor.onDatabaseOpened(db)
        syncStatusProvider.onDatabaseOpened(db, openResult)
        observerBus.notifyDatabaseOpened(db)
    }

    private fun onDatabaseClosed() {
        lockInteractor.onDatabaseClosed()
        syncStatusProvider.onDatabaseClosed()
        observerBus.notifyDatabaseClosed()
    }

    private fun openDatabase(
        type: KeepassImplementation,
        fsResolver: FileSystemResolver,
        fsOptions: FSOptions,
        file: FileDescriptor,
        input: OperationResult<InputStream>,
        key: EncryptedDatabaseKey
    ): OperationResult<EncryptedDatabase> {
        val openResult = when (type) {
            KeepassImplementation.KOTPASS -> KotpassDatabase.open(
                fsResolver,
                fsOptions,
                file,
                input,
                key
            )
            KeepassImplementation.KEEPASS_RS -> KeepassRsDatabase.open(
                fsResolver,
                fsOptions,
                file,
                input,
                key
            ).toOperationResult()
        }

        if (openResult.isFailed) {
            return openResult.takeError()
        }

        val db = openResult.getOrThrow()
            .apply {
                this.watcher.subscribe(syncStatusProvider)
            }

        return openResult.takeStatusWith(db)
    }

    private data class DatabaseReference(
        val type: KeepassImplementation,
        val database: EncryptedDatabase
    )
}